src/shadowbox/server/server_access_key.ts

```diff
@@ -90,75 +139,178 @@ export class ServerAccessKeyRepository implements AccessKeyRepository {
+    return accessKey ? accessKey.metricsId : undefined;
+  }
+
+  // Compares access key usage with collected metrics, marking them as under or over quota.
+  async enforceAccessKeyQuotas() {
+    let quotaStatusChanged = false;
+    for (const accessKey of this.accessKeys) {
+      quotaStatusChanged = quotaStatusChanged || await this.updateAccessKeyQuotaStatus(accessKey);
+    }
+    if (quotaStatusChanged) {
+      this.updateServer();
+    }
+  }
+
+  // Updates `accessKey` quota status by comparing its usage with collected metrics. Returns whether
+  // the quota status changed.
+  private async updateAccessKeyQuotaStatus(accessKey: ServerAccessKey): Promise<boolean> {
+    if (!accessKey.quota) {
+      return false;  // Don't query the usage of access keys without quota.
+    }
+    const usageBytes =
+        await this.metrics.getOutboundByteTransfer(accessKey.id, accessKey.quota.windowHours);
+    const isOverQuota = usageBytes > accessKey.quota.quotaBytes;
+    logging.debug(`Enforcing quota for access key ${accessKey.id}. Quota: ${
+        JSON.stringify(accessKey.quota)}, usage: ${usageBytes}B, isOverQuota: ${isOverQuota}`);
+    const quotaStatusChanged = isOverQuota !== accessKey.isOverQuota;
+    accessKey.isOverQuota = isOverQuota;
+    return quotaStatusChanged;
   }
 
   private updateServer(): Promise<void> {
-    return this.shadowsocksServer.update(this.keyConfig.data().accessKeys.map((e) => {
-      return {id: e.id, port: e.port, cipher: e.encryptionMethod, secret: e.password};
-    }));
+    const serverAccessKeys = this.accessKeys.filter(key => !key.isOverQuota).map(key => {
+      return {
+        id: key.id,
+        port: key.proxyParams.portNumber,
+        cipher: key.proxyParams.encryptionMethod,
+        secret: key.proxyParams.password
+      };
+    });
+    return this.shadowsocksServer.update(serverAccessKeys);
+  }
+
+  private loadAccessKeys(): AccessKey[] {
+    return this.keyConfig.data().accessKeys.map(key => makeAccessKey(this.proxyHostname, key));
+  }
+
+  private saveAccessKeys() {
+    try {
+      this.keyConfig.data().accessKeys = this.accessKeys.map(key => makeAccessKeyJson(key));
+      this.keyConfig.write();
+    } catch (error) {
+      throw new Error(`Failed to save access key config: ${error}`);
+    }
   }
 
-  private getAccessKey(id: AccessKeyId): AccessKeyConfig {
-    for (const accessKeyJson of this.keyConfig.data().accessKeys) {
-      if (accessKeyJson.id === id) {
-        return accessKeyJson;
+  // Returns a reference to the access key with `id`, or undefined if the key is not found.
+  private getAccessKey(id: AccessKeyId): ServerAccessKey|undefined {
+    for (const accessKey of this.accessKeys) {
+      if (accessKey.id === id) {
+        return accessKey;
       }
     }
     return undefined;
   }
 }
+
+// Retrieves access key usage metrics from a Prometheus instance.
+export class AccessKeyUsageMetrics implements UsageMetrics {
```

Comment:
'Usage', 'ByteTransfer', 'Quota Traffic'. Can we use a consistent phrase?

Maybe `getQuotaTrafficBytes(...)` and `AccessKeyQuotaTraffic implements QuotaTraffic`?

Reply:
Done in 20d622d. Went with 'byte transfer' for descriptiveness. This is consistent with the quota feature documentation, as well as with other metrics-related terminology.